relay-kafka/src/producer.rs

```diff
@@ -0,0 +1,170 @@
+use std::{collections::BTreeMap, sync::Arc};
+
+use rdkafka::{producer::BaseRecord, ClientConfig};
+
+use relay_statsd::metric;
+
+use crate::{
+    config::{KafkaConfig, KafkaParams},
+    utils::{CaptureErrorContext, KafkaHistograms, ThreadedProducer},
+    Message, ProducerError,
+};
+
+/// Temporary map used to deduplicate kafka producers
+type ReusedProducersMap<'a> = BTreeMap<Option<&'a str>, Arc<ThreadedProducer>>;
+
+/// This object containes the Kafka producer variants for single and sharded configurations.
+pub enum Producer {
+    Single {
+        topic_name: String,
+        producer: Arc<ThreadedProducer>,
+    },
+    Sharded(ShardedProducer),
+}
+
+impl Producer {
+    /// Creates the kafka [`Producer`] based on the provided configuration.
+    ///
+    /// # Errors
+    /// Returns [`ProducerError::InvalidConfig`] error if the provided configuration is wrong and
+    /// the producer could not be created.
+    pub fn create<'a>(
+        config: &'a KafkaConfig,
+        reused_producers: &mut ReusedProducersMap<'a>,
```

Comment:
This was fine while this was still in a private module, but this mixes several concerns for a public signature:
- It creates a new producer instance
- It memoizes them and determines the data structure for memoization

I realize that this increases the effort for this refactor, but it would be good to clean this up and separate those concerns. Ideally the memo-map is either generic, or becomes a higher-level utility that uses `create` internally.

In the end, I think the abstraction here is entirely wrong. The real producer is the `Arc`'ed threaded producer inside. The `Producer` type here then gets one or more of them and wraps them together with a topic name.

Reply:
I will revisit this and make some changes into this refactor.